src/allocator.rs

```diff
@@ -332,13 +336,36 @@ impl Allocator {
 
     pub fn new_pair(&mut self, first: NodePtr, rest: NodePtr) -> Result<NodePtr, EvalErr> {
         let idx = self.pair_vec.len();
-        if idx == MAX_NUM_PAIRS {
+        if idx >= self.max_num_pairs {
             return err(self.nil(), "too many pairs");
         }
         self.pair_vec.push(IntPair { first, rest });
         Ok(NodePtr::new(ObjectType::Pair, idx))
     }
 
+    pub fn reduce_pair_max(&mut self, amount: usize) -> Result<(), EvalErr> {
+        if self.max_num_pairs - amount <= self.pair_vec.len() {
+            return err(
+                self.nil(),
+                format!(
+                    "reduce_pair_max: amount exceeds current pair count ({} pairs allocated, {} max pairs)",
+                    self.pair_vec.len(),
+                    self.max_num_pairs
+                ).as_str(),
+            );
+        }
+        self.max_num_pairs -= amount;
+        Ok(())
+    }
+
+    pub fn increase_pair_max(&mut self, amount: usize) -> Result<(), EvalErr> {
+        // increase the max number of pairs by the amount, but don't exceed usize::MAX
+        let max_increase = usize::MAX - self.max_num_pairs;
+        let increase = amount.min(max_increase);
+        self.max_num_pairs += increase;
```

Comment:
```suggestion
        self.max_num_pairs = u32::.saturating_add(self.max_num_paurs, increase);
```

Reply:
done with `self.max_num_pairs.saturating_add(amount)`

Reply:
but there's also a risk that the `+` will exceed the capacity of the variable. Perhaps you only ever call this with `amount=1`, in which case it's safe. But in that case, maybe this function should always increment by 1.